site/docs/tutorial/sql_to_substrait.md

```diff
@@ -0,0 +1,1305 @@
+  product_name,
+  product_id
+```
+
+The query asked the question: For products that aren't in the `"Computer"`
+category, how much has each product generated in sales?
+
+However, Substrait doesn't correspond to SQL as much as it does to logical plans.
+So to be less ambiguous, the plan we are aiming for looks like:
+
+```
+|-+ Project(products.product_name, orders.product_id, sales)
+  |-+ Aggregate({sales = sum(quantity_price)}, group_by=(product_name, product_id))
+    |-+ InnerJoin(on=orders.product_id = products.product_id)
+      |- ReadTable(orders)
+      |-+ Filter(INDEX_IN("Computers", categories) IS NULL)
+        |- ReadTable(products)
+```
+
+## Types and Schemas
+
+As part of the Substrait plan, we'll need to embed the data types of the input
+tables. In Substrait, each type is a distinct message, which at a minimum 
+contains a field for nullability. For example, a string field looks like:
+
+```json
+{
+  "string": {
+    "nullability": "NULLABILITY_NULLABLE"
+  }
+}
+```
+
+Nullability is an enum not a boolean, since Substrait allows 
+`NULLABILITY_UNSPECIFIED` as an option.
+
+Other types such as `VarChar` and `Decimal` have other parameters. For example,
+our `orders.price` column will be represented as:
+
+```json
+{
+  "decimal": {
+    "scale": 10,
+    "precision": 2,
+    "nullability": "NULLABILITY_NULLABLE"
+  }
+}
+```
+
+Finally, there are nested compound types such as structs and list types that 
+have other types as parameters. For example, the `products.categories` column
+is a list of strings, so can be represented as:
+
+```json
+{
+  "list": {
+    "type": {
+      "string": {
+        "nullability": "NULLABILITY_REQUIRED"
+      }
+    },
+    "nullability": "NULLABILITY_REQUIRED"
+  }
+}
+```
+
+Schemas of tables can be represented with a `NamedStruct` message, which is the
+combination of a struct type containing all the columns and a list of column
+names. For the `orders` table, this will look like:
+
+```json
+{
+  "names": [
+    "product_id",
+    "quantity",
+    "order_date",
+    "price"
+  ],
+  "struct": {
+    "types": [
+      {
+        "i64": {
+          "nullability": "NULLABILITY_REQUIRED"
+        }
+      },
+      {
+        "i32": {
+          "nullability": "NULLABILITY_REQUIRED"
+        }
+      },
+      {
+        "date": {
+          "nullability": "NULLABILITY_REQUIRED"
+        }
+      },
+      {
+        "decimal": {
+          "scale": 10,
+          "precision": 2,
+          "nullability": "NULLABILITY_NULLABLE"
+        }
+      }
+    ],
+    "nullability": "NULLABILITY_REQUIRED"
+  }
+}
+```
+
+Here, `names` is the names of all fields. In nested schemas, this includes the
+names of subfields in depth-first order. So for the `products` table, the 
+`details` struct field will be included as well as the two subfields (`manufacturer`
+and `year_created`) right after. And because it's depth first, these subfields
+appear before `product_name`. The full schema looks like:
```

Comment:
Is this correct? There is no subfield for list or map types? (categories here is a list of strings)

Reply:
I think so but I'm not certain.  I don't think a list column has names (even if it is a fixed size list) for its children.  In a map column the names are part of the data and not the field.